### PR TITLE
Upgrade agent changes

### PR DESCRIPTION
**Pull request by bernardjkim**
## Description

This PR makes a few improvements to the gravity upgrade agents. 
- The gravity-agents will now run as regular services instead of oneshot services so that they restart on failure and node reboot. 
- A `gravity agent status` command is now available to view the status of the deployed gravity agents. Will look something like:
```
// Successfully received status response from all agents
Tue Jul 14 07:11:19 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-2       172.28.128.102     Deployed     7.1.0-alpha.1.141
node-3       172.28.128.103     Deployed     7.1.0-alpha.1.141
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.141

// Failed to receive status response from all agents
Tue Jul 14 07:12:30 UTC Collecting RPC agent status
Hostname     Address            Status      Version
--------     -------            ------      -------
node-2       172.28.128.102     Offline
node-3       172.28.128.103     Offline
node-1       172.28.128.101     Offline
```

- The `gravity agent deploy` command now supports a `--version` flag to specify a particular agent version to deploy.
- The `gravity agent deploy` command also accepts a node hostname as an argument to deploy an agent on the specified node.
## Type of change

* New feature (non-breaking change which adds functionality)

## Linked tickets and other PRs

* Refs https://github.com/gravitational/gravity/issues/1667
* Requires https://github.com/gravitational/gravity.e/pull/4321

## TODOs

- [x] Self-review the change
- [x] Perform manual testing
- [x] Address review feedback

## Testing done

**Verify gravity-agent is running after reboot**

**Verify lead update agent does not automatically attempt to resume upgrade**

**Verify upgrade can be resumed after node reboot with `gravity upgrade --resume`**

**Verify all gravity-agents are shutdown after upgrade**

**Verify gravity-agents are disabled and do not restart on next reboot**

**Verify upgrade can be resumed after non lead agent restarts**

**Verify upgrade after rollback**

**Deploy particular agent version with --version flag**
```
[vagrant@node-1 installer]$ sudo gravity agent deploy --version=7.1.0-alpha.1.153
Wed Jul 29 19:13:08 UTC Deploying agents on the cluster nodes
Deployed agent on node-3 (172.28.128.103)
Deployed agent on node-1 (172.28.128.101)
Deployed agent on node-2 (172.28.128.102)

[vagrant@node-1 installer]$ sudo gravity agent status
Wed Jul 29 19:13:31 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-3       172.28.128.103     Deployed     7.1.0-alpha.1.153
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.153
node-2       172.28.128.102     Deployed     7.1.0-alpha.1.153
```

**Deploy onto a particular node**
```
[vagrant@node-1 installer]$ sudo gravity agent deploy node-1
Wed Jul 29 19:00:36 UTC Deploying agents on the cluster nodes
Deployed agent on node-1 (172.28.128.101)

[vagrant@node-1 installer]$ sudo gravity agent status
Wed Jul 29 19:00:43 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.154
node-3       172.28.128.103     Offline
node-2       172.28.128.102     Offline
```

**Verify agents show deployed even if older server version**
```
[vagrant@node-1 ~]$ sudo gravity agent status
Wed Jul 29 21:22:44 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-2       172.28.128.102     Deployed     N/A
node-3       172.28.128.103     Deployed     N/A
node-1       172.28.128.101     Deployed     N/A
```

**Verify deploy without args works as usual**
```
[vagrant@node-1 ~]$ sudo gravity agent deploy
Thu Jul 30 21:21:04 UTC Deploying agents on the cluster nodes
Deployed agent on node-2 (172.28.128.102)
Deployed agent on node-3 (172.28.128.103)
Deployed agent on node-1 (172.28.128.101)

[vagrant@node-1 ~]$ sudo gravity agent status
Thu Jul 30 21:21:12 UTC Collecting RPC agent status
Hostname     Address            Status       Version
--------     -------            ------       -------
node-2       172.28.128.102     Deployed     7.1.0-alpha.1.162
node-3       172.28.128.103     Deployed     7.1.0-alpha.1.162
node-1       172.28.128.101     Deployed     7.1.0-alpha.1.162
```

### CODE REVIEW

File: tool/gravity/cli/rpcagent.go [hunk 1]
```diff
@@ -366,6 +367,30 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
 	return trace.Wrap(err)
 }
 
+// rpcAgentStatus requests and writes the gravity agent statuses to stdout.
+func rpcAgentStatus(env *localenv.LocalEnvironment) error {
```
Comment (r0mant) [inline tool/gravity/cli/rpcagent.go:371]: How does this method behave / what does it output if there are no deployed agents?

File: lib/update/cluster/automatic.go [hunk 2]
```diff
@@ -132,3 +132,14 @@ func ShutdownClusterAgents(ctx context.Context, remote rpc.AgentRepository) erro
 	err = rpc.ShutdownAgents(ctx, nodes, log.StandardLogger(), remote)
 	return trace.Wrap(err)
 }
+
+// AgentStatus requests the statuses of all agents
+func AgentStatus(ctx context.Context, remote rpc.AgentRepository) ([]rpc.GravityAgentStatus, error) {
+	nodes, err := kubectl.GetNodesAddr(ctx)
```
Comment (r0mant) [inline lib/update/cluster/automatic.go:138]: I think we should retrieve the nodes from the cluster state, not k8s. If you pass an operator here, you can use it:

```go
// in cli/rpcagent.go
func rpcAgentStatus(env *localenv.LocalEnvironment) error {
  operator, err := env.ClusterOperator()
  ...
  cluster.AgentStatus(ctx, operator, remote)
}

// here
func AgentStatus(ctx context.Context, operator ops.Operator, remote ...) () {
  cluster, err := operator.GetLocalSite()
  ...
  cluster.ClusterState.Servers // <-- cluster nodes
}
```

File: lib/update/cluster/automatic.go [hunk 3]
```diff
@@ -132,3 +132,14 @@ func ShutdownClusterAgents(ctx context.Context, remote rpc.AgentRepository) erro
 	err = rpc.ShutdownAgents(ctx, nodes, log.StandardLogger(), remote)
 	return trace.Wrap(err)
 }
+
+// AgentStatus requests the statuses of all agents
+func AgentStatus(ctx context.Context, remote rpc.AgentRepository) ([]rpc.GravityAgentStatus, error) {
```
Comment (r0mant) [inline lib/update/cluster/automatic.go:137]: I think this better belongs in `lib/rpc`?

File: tool/gravity/cli/rpcagent.go (hunk 1 above)
Comment (r0mant) [reply tool/gravity/cli/rpcagent.go:371]: Also, what does it output if there's a failure contacting one of the agents / retrieving info from it?

File: lib/rpc/status.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/gravitational/gravity/lib/modules"
+	"github.com/gravitational/gravity/lib/storage"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// GravityAgentStatus contains a gravity agent's status information.
+type GravityAgentStatus struct {
```
Comment (r0mant) [inline lib/rpc/status.go:33]: Nit: Just `AgentStatus`, or `RPCAgentStatus`.

File: lib/rpc/status.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	Node string
+	// IP specifies the IP address of the node running the agent.
+	IP string
+	// Status indiciates the current status of the agent. An agent is `Deployed`
+	// if the gravity-agent service is active. The agent is `Offline` if it
+	// fails to respond to the status request.
+	Status string
+	// Version describes gravity agent version.
+	Version string
+}
```
Comment (r0mant) [inline lib/rpc/status.go:44]: I would add another field here with `Error error` just so it can be displayed in case there's a failure contacting the agent.

File: lib/rpc/status.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+func AgentStatus(ctx context.Context, servers []string, logger log.FieldLogger, rpc AgentRepository) (
+	[]GravityAgentStatus, error) {
+	errs := make(chan error, len(servers))
+	statusCh := make(chan GravityAgentStatus, len(servers))
+	for _, srv := range servers {
+		go func(host string) {
+			systemInfo, err := getSystemInfo(ctx, host, rpc)
```
Comment (r0mant) [inline lib/rpc/status.go:52]: I would encapsulate these two (getSystemInfo and getVersion) into a separate function (getAgentStatus) just so the error handling is a little easier, for example:

```go
func getAgentStatus(ctx context.Context, address string, rpc AgentRepository) AgentStatus {
  systemInfo, err := getSystemInfo(ctx, host, rpc)
  // return status with Error set
  version, err := getVersion(ctx, host, rpc)
  // return status with Error set
  return AgentStatus{
    Address:  address,
    Status:   constants.GravityAgentDeployed,
    Hostname: systemInfo.GetHostname(),
    Version:  version.Version,    
  }, nil
}

// Then the goroutine becomes trivial:

go func(address string) {
  statusCh <- getAgentStatus(ctx, address, rpc)
}
```

File: lib/rpc/status.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// GravityAgentStatus contains a gravity agent's status information.
+type GravityAgentStatus struct {
+	// Node specifies the name of the node running the agent.
+	Node string
+	// IP specifies the IP address of the node running the agent.
+	IP string
```
Comment (r0mant) [inline lib/rpc/status.go:37]: Nit: I'd rename `IP -> Address` and `Node -> Hostname` to be consistent with naming in other places.

**Review by r0mant — COMMENTED** [review]
Preliminary feedback for your draft.

File: lib/rpc/status.go (hunk 6 above)
Comment (bernardjkim) [reply lib/rpc/status.go]: Realized we don't even need to request systemInfo since we are only collecting hostname from it. We can pull hostname from the cluster state.

File: tool/gravity/cli/rpcagent.go (hunk 1 above)
Comment (bernardjkim) [reply tool/gravity/cli/rpcagent.go:376]: Added additional details and added a 5 sec timeout on the status request.
If an agent fail to return a status response, the agent will be displayed as `Offline` and the version column will be empty.

File: lib/constants/constants.go [hunk 8]
```diff
@@ -230,6 +230,12 @@ const (
 	// OneshotService is a service that executes one time
 	OneshotService = "oneshot"
 
+	// SimpleService is a simple service that is recommended for long running processes
```
Comment (a-palchikov) [inline lib/constants/constants.go:233]: I think a link to the systemd docs as a reference to service types would be useful.
Also, I think these could be more appropriate in [lib/system/service](https://github.com/gravitational/gravity/tree/bernard/master/upgrade-agent/lib/system/service) but it's debatable.

File: lib/rpc/proto/discovery.proto [hunk 9]
```diff
@@ -16,6 +16,15 @@ service Discovery {
...
     rpc GetCurrentTime(google.protobuf.Empty) returns (google.protobuf.Timestamp);
+
+    // GetVersion returns the agent's version information.
+    rpc GetVersion(google.protobuf.Empty) returns (Version);
+}
+
+// Version describes an agent's version information
+message Version {
+    // Payload is the version information payload.
+    bytes payload = 1;
```
Comment (a-palchikov) [inline lib/rpc/proto/discovery.proto:27]: I would rather have the protobuf to be the source of truth for data types.

File: lib/rpc/status.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+	Hostname string
+	// Address specifies the IP address of the node running the agent.
+	Address string
+	// Status indiciates the current status of the agent. An agent is `Deployed`
+	// if the gravity-agent service is active. The agent is `Offline` if it
+	// fails to respond to the status request.
+	Status string
+	// Version describes gravity agent version.
+	Version string
+	// Error contains errors while collected while requesting agent status.
```
Comment (a-palchikov) [inline lib/rpc/status.go:42]: nit: maybe rephrase this to avoid duplicate `while`: `Error contains an error that might have occurred when requesting agent status`

File: lib/rpc/status.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/gravitational/trace"
+)
+
+// AgentStatus contains a gravity agent's status information.
+type AgentStatus struct {
+	// Hostname specifies the hostname of the node running the agent.
+	Hostname string
+	// Address specifies the IP address of the node running the agent.
+	Address string
+	// Status indiciates the current status of the agent. An agent is `Deployed`
```
Comment (a-palchikov) [inline lib/rpc/status.go:36]: indicates

File: lib/rpc/status.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+		return agentStatus
+	}
+
+	agentStatus.Version = version.Version
+	agentStatus.Status = constants.GravityAgentDeployed
+	return agentStatus
+}
+
+func getVersion(ctx context.Context, addr string, rpc AgentRepository) (version modules.Version, err error) {
+	ctx, cancel := context.WithTimeout(ctx, defaults.DialTimeout)
```
Comment (a-palchikov) [inline lib/rpc/status.go:82]: This sounds like a bug - `DialTimeout` should only affect the connection step and not the actual API call.

File: tool/gravity/cli/system.go [hunk 13]
```diff
@@ -280,6 +282,33 @@ func reinstallOneshotService(env *localenv.LocalEnvironment, serviceName string,
 	return trace.Wrap(err)
 }
 
+// reinstallService stops and reinstalls the service specified by serviceName
+// as a simple service.
+func reinstallService(env *localenv.LocalEnvironment, serviceName string, cmd []string) error {
```
Comment (a-palchikov) [inline tool/gravity/cli/system.go:287]: I'd really like that we moved general purpose functions to specific packages - i.e. `lib/system/service` if there's no such API yet.

File: lib/defaults/defaults.go [hunk 14]
```diff
@@ -219,6 +219,9 @@ const (
 	// AgentStopTimeout is amount of time agent gets to gracefully shut down
 	AgentStopTimeout = 10 * time.Second
 
+	// AgentStatusTimeout specifies the timeout for agent status query
+	AgentStatusTimeout = 5 * time.Second
```
Comment (a-palchikov) [inline lib/defaults/defaults.go:223]: I'm sort of on the fence with timeouts - I feel these should be configurable via envars at least to avoid any issues with slower networks and arbitrary cluster sizes.

Comment (bernardjkim) [reply lib/defaults/defaults.go:223]: Gravity will now check for `GRAVITY_AGENT_STATUS_TIMEOUT` env var. Will default to 5sec if env var is not specified.

File: lib/rpc/proto/discovery.proto (hunk 9 above)
Comment (bernardjkim) [reply lib/rpc/proto/discovery.proto]: Moved Version definition from modules to proto pkg. a583350

File: lib/update/cluster/automatic.go [hunk 15]
```diff
@@ -43,6 +43,21 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
...
+
+	plan, err := fsm.GetOperationPlan(updateEnv.Backend, siteOp.Key())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Check if the upgrade is already in progress, which would indicate an
+	// agent is already running. Return without creating a new agent.
+	for _, phase := range plan.Phases {
+		if phase.IsInProgress() {
```
Comment (r0mant) [inline lib/update/cluster/automatic.go:56]: I think we should be checking for "not unstarted" phases. Otherwise, this will not catch scenarios where a phase is "failed" or "completed". So we should basically look for phases in any state other than the initial "unstarted" one.

File: lib/update/cluster/automatic.go [hunk 16]
```diff
@@ -43,6 +43,21 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
...
+			return nil
```
Comment (r0mant) [inline lib/update/cluster/automatic.go:57]: We should definitely log something in this case.

File: lib/utils/env.go [hunk 17]
```diff
@@ -131,6 +132,19 @@ func GetenvInt(name string) (int, error) {
 	return valueI, nil
 }
 
+//GetenvDuration returns the specified environment variable value parsed as a duration.
```
Comment (r0mant) [inline lib/utils/env.go:135]: Nit: Space after //.

File: tool/gravity/cli/rpcagent.go [hunk 18]
```diff
@@ -315,16 +331,44 @@ func deployAgents(ctx context.Context, env *localenv.LocalEnvironment, req deplo
...
+		verifiedServers, err := verifyCluster(ctx, req)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		servers = append(servers, verifiedServers...)
+	}
+
+	// If version is not specified in the request, use the current build version
+	if req.version == "" {
+		req.version = strings.Split(version.Get().Version, "+")[0]
```
Comment (r0mant) [inline tool/gravity/cli/rpcagent.go:358]: What's "split by +" for?

File: tool/gravity/cli/rpcagent.go [hunk 19]
```diff
@@ -388,6 +443,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
...
+
+	t := goterm.NewTable(0, 10, 5, ' ', 0)
+	common.PrintTableHeader(t, []string{"Hostname", "Address", "Status", "Version"})
+	for _, status := range statusList {
+		fmt.Fprintf(t, "%s\t%s\t%s\t%s\n", status.Hostname, status.Address, status.Status, status.Version)
+		if status.Error != nil {
+			log.WithError(err).Debugf("Failed to collect query agent version on %s.", status.Address)
+		}
+	}
+	fmt.Println(t.String())
```
Comment (r0mant) [inline tool/gravity/cli/rpcagent.go:486]: We usually use `env.Println` - it considers some flags potentially passed on the CLI, like "quiet" or "verbose" that might affect the output.

File: tool/gravity/cli/rpcagent.go [hunk 20]
```diff
@@ -388,6 +443,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
...
+	defer cancel()
+
+	statusList := rpc.CollectAgentStatus(ctx, cluster.ClusterState.Servers, fsm.NewAgentRunner(creds))
+
+	t := goterm.NewTable(0, 10, 5, ' ', 0)
+	common.PrintTableHeader(t, []string{"Hostname", "Address", "Status", "Version"})
+	for _, status := range statusList {
+		fmt.Fprintf(t, "%s\t%s\t%s\t%s\n", status.Hostname, status.Address, status.Status, status.Version)
+		if status.Error != nil {
+			log.WithError(err).Debugf("Failed to collect query agent version on %s.", status.Address)
```
Comment (r0mant) [inline tool/gravity/cli/rpcagent.go:483]: Shouldn't this be `log.WithError(status.Error)`?

File: tool/gravity/cli/rpcagent.go [hunk 21]
```diff
@@ -388,6 +443,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
...
+		}
+	}
+	fmt.Println(t.String())
+
+	return trace.Wrap(err)
```
Comment (r0mant) [inline tool/gravity/cli/rpcagent.go:488]: It's a bit weird - what is this error that's returned here? The duration parsing error from above?

File: tool/gravity/cli/rpcagent.go [hunk 22]
```diff
@@ -345,6 +389,17 @@ func newDeployAgentsRequest(ctx context.Context, env *localenv.LocalEnvironment,
 	}, nil
 }
 
+// findServerByHostname returns the server with the matching hostname. Returns
+// a NotFound error if a server with a matching hostname does not exist.
+func findServerByHostname(clusterState storage.ClusterState, hostname string) (server storage.Server, err error) {
```
Comment (r0mant) [inline tool/gravity/cli/rpcagent.go:394]: I'm pretty sure we already have this function in storage.go.

https://github.com/gravitational/gravity/blob/master/lib/storage/storage.go#L696

File: lib/rpc/status.go [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+func getAgentStatus(ctx context.Context, server storage.Server, rpc AgentRepository) AgentStatus {
+	agentStatus := AgentStatus{
+		Hostname: server.Hostname,
+		Address:  server.AdvertiseIP,
+		Status:   constants.GravityAgentOffline,
+	}
+
+	version, err := getVersion(ctx, server.AdvertiseIP, rpc)
+	if err != nil {
```
Comment (r0mant) [inline lib/rpc/status.go:71]: I think this should also consider a "not implemented" error, in case it contacts the agent that doesn't have GetVersion method yet. In this case the agent is basically online, we just don't know its version (could display it as `N/A` for example).

File: tool/gravity/cli/rpcagent.go (hunk 18 above)
Comment (bernardjkim) [reply tool/gravity/cli/rpcagent.go]: https://github.com/gravitational/gravity/blob/46b9ee065b121ed99ac593cb98084ef079ac020b/tool/gravity/cli/rpcagent.go#L416
Copied over from `getGravityPackage` func. It doesn't seem like we need to split version by '+'. I guess we had a different version format in the past?

File: lib/rpc/status.go [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+		go func(server storage.Server) {
+			statusCh <- getAgentStatus(ctx, server, rpc)
+		}(srv)
+	}
+
+	var statusList []AgentStatus
+	for range servers {
+		statusList = append(statusList, <-statusCh)
+	}
+
```
Comment (a-palchikov) [inline lib/rpc/status.go:59]: nit: could also close the channel at this point to help gc.

File: lib/update/cluster/automatic.go [hunk 25]
```diff
@@ -43,6 +43,22 @@ func AutomaticUpgrade(ctx context.Context, localEnv, updateEnv *localenv.LocalEn
...
+	siteOp := (*ops.SiteOperation)(operation)
+
+	plan, err := fsm.GetOperationPlan(updateEnv.Backend, siteOp.Key())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Check if the upgrade is already in progress, which would indicate an
+	// agent is already running. Return without creating a new agent.
+	for _, phase := range plan.Phases {
```
Comment (a-palchikov) [inline lib/update/cluster/automatic.go:55]: This should be better encapsulated as `IsOperationInProgress` or `IsPlanInProgress` - you would not need the comment then. I'd check whether we might already have an API like that.

File: tool/gravity/cli/checks.go [hunk 26]
```diff
@@ -134,7 +134,7 @@ func checkUpgrade(ctx context.Context, env *localenv.LocalEnvironment, config pr
 		return trace.Wrap(err)
 	}
 	// Deploy RPC agents that will be used for running checks on the nodes.
-	credentials, err := rpcAgentDeployHelper(ctx, env, "", "")
+	credentials, err := rpcAgentDeployHelper(ctx, env, "", "", "", "")
```
Comment (a-palchikov) [inline tool/gravity/cli/checks.go:137]: Pretty hard to tell what the values are and why empty. In this case I'd suggest an options struct which would be provided to the function as empty.

File: tool/gravity/cli/rpcagent.go [hunk 27]
```diff
@@ -388,6 +432,51 @@ func rpcAgentShutdown(env *localenv.LocalEnvironment) error {
...
+	defer cancel()
+
+	statusList := rpc.CollectAgentStatus(ctx, cluster.ClusterState.Servers, fsm.NewAgentRunner(creds))
+
+	t := goterm.NewTable(0, 10, 5, ' ', 0)
+	common.PrintTableHeader(t, []string{"Hostname", "Address", "Status", "Version"})
+	for _, status := range statusList {
+		fmt.Fprintf(t, "%s\t%s\t%s\t%s\n", status.Hostname, status.Address, status.Status, status.Version)
+		if status.Error != nil {
+			log.WithError(status.Error).Debugf("Failed to collect agent status on %s.", status.Address)
```
Comment (a-palchikov) [inline tool/gravity/cli/rpcagent.go:472]: I wonder if we need to indicate with that command whether (any) agents are offline via a non-0 exit code?
Also, if used with `--quiet` the command will not output anything - not sure if this is intentional/desired.

File: tool/gravity/cli/rpcagent.go [hunk 28]
```diff
@@ -315,16 +331,44 @@ func deployAgents(ctx context.Context, env *localenv.LocalEnvironment, req deplo
 
 // newDeployAgentsRequest creates a new request to deploy agents on the local cluster
 func newDeployAgentsRequest(ctx context.Context, env *localenv.LocalEnvironment, req deployAgentsRequest) (*rpc.DeployAgentsRequest, error) {
-	servers, err := verifyCluster(ctx, req)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	// If hostname is specified in the request, deploy agent only on specified node
+	var servers []rpc.DeployServer
+	if req.hostname != "" {
```
Comment (a-palchikov) [inline tool/gravity/cli/rpcagent.go:336]: Not sure about the need to leak hostname into deployAgentRequest - if it's desired that only an agent on a specific node is started, add an optional server (or a server slice - a single element means only on a single node) to the request.

File: tool/gravity/cli/rpcagent.go (hunk 27 above)
Comment (bernardjkim) [reply tool/gravity/cli/rpcagent.go:481]: It could be useful to have the command return a non-0 exit code if any agents are offline. Updated in cd7ee63.

I think it was preferred to use env.Println. Not entirely sure when someone would want to run the agent status command without any output, but since it now returns an error if any agents are offline, it might be useful somewhere.

File: lib/system/service/simple.go [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+// specified serviceName.
+func ReinstallSimpleService(serviceName string, cmd []string) error {
+	services, err := systemservice.New()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = services.StopService(serviceName)
+	if err != nil {
+		logrus.Warnf("Error stopping service %v: %v.", serviceName, err)
```
Comment (r0mant) [inline lib/system/service/simple.go:40]: `logrus.WithError(err).Warnf()`

**Review by r0mant — APPROVED** [review]
Code looks good, one additional suggestion for "testing done" is in addition to `Verify all gravity-agents are shutdown after upgrade` scenario, let's also make sure that the shutdown agents are not automatically restarted after reboot (i.e. complete the upgrade, reboot the node(-s) and make sure agents aren't running).

File: tool/gravity/cli/rpcagent.go (hunk 18 above)
Comment (r0mant) [reply tool/gravity/cli/rpcagent.go]: @a-palchikov Do you recall by any chance why we had this "split by +"? I remember it was needed for something but can't remember why.

**Review by a-palchikov — APPROVED** [review]

File: tool/gravity/cli/rpcagent.go (hunk 18 above)
Comment (a-palchikov) [reply tool/gravity/cli/rpcagent.go]: It might have had something to do with semver parsing but I need to go through the history to know more/remember.
